packages/datadog-instrumentations/src/mongoose.js

```diff
@@ -26,5 +26,147 @@ addHook({
+  versions: ['>=4.6.4 <5', '5', '6', '>=7'],
+  file: 'lib/model.js'
+}, Model => {
+  [...collectionMethodsWithFilter, ...collectionMethodsWithTwoFilters].forEach(methodName => {
+    const useTwoArguments = collectionMethodsWithTwoFilters.includes(methodName)
+
+    try {
+      shimmer.wrap(Model, methodName, method => {
+        return function wrappedModelMethod () {
+          if (!startCh.hasSubscribers) {
+            return method.apply(this, arguments)
+          }
+
+          const asyncResource = new AsyncResource('bound-anonymous-fn')
+
+          const filters = [arguments[0]]
+          if (useTwoArguments) {
+            filters.push(arguments[1])
+          }
+
+          const finish = asyncResource.bind(function () {
+            finishCh.publish()
+          })
+
+          let callbackWrapped = false
+          const lastArgumentIndex = arguments.length - 1
+
+          if (typeof arguments[lastArgumentIndex] === 'function') {
+            // is a callback, wrap it to execute finish()
+            shimmer.wrap(arguments, lastArgumentIndex, originalCb => {
+              return function () {
+                finish()
+
+                return originalCb.apply(this, arguments)
+              }
+            })
+
+            callbackWrapped = true
+          }
+
+          return asyncResource.runInAsyncScope(() => {
+            startCh.publish({
+              filters,
+              methodName
+            })
+
+            const res = method.apply(this, arguments)
+
+            // if it is not callback, wrap exec method and its then
+            if (!callbackWrapped) {
+              shimmer.wrap(res, 'exec', originalExec => {
+                return function wrappedExec () {
+                  const execResult = originalExec.apply(this, arguments)
+
+                  // wrap them method, wrap resolve and reject methods
+                  shimmer.wrap(execResult, 'then', originalThen => {
+                    return function wrappedThen () {
+                      const resolve = arguments[0]
+                      const reject = arguments[1]
+
+                      // not using shimmer here because resolve/reject could be empty
+                      arguments[0] = function wrappedResolve () {
+                        finish()
+
+                        if (resolve) {
+                          return resolve.apply(this, arguments)
+                        }
+                      }
+
+                      arguments[1] = function wrappedReject () {
+                        finish()
+
+                        if (reject) {
+                          return reject.apply(this, arguments)
+                        }
+                      }
+
+                      return originalThen.apply(this, arguments)
+                    }
+                  })
+
+                  return execResult
+                }
+              })
+            }
+            return res
+          })
+        }
+      })
+    } catch (e) {
```

Comment:
Error swallowing is usually pretty dangerous. What about preceding the wrap line with a conditional:

```js
if (!(methodName in Model)) return
shimmer.wrap(Model, methodName, method => {
```